src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php

```diff
@@ -0,0 +1,259 @@
+        }
+
+        return $configuration;
+    }
+
+    /**
+     * @param int $delay The delay in milliseconds
+     *
+     * @throws \Doctrine\DBAL\DBALException
+     */
+    public function send(string $body, array $headers, int $delay = 0): void
+    {
+        $now = (\DateTime::createFromFormat('U.u', microtime(true)));
+        $availableAt = (clone $now)->modify(sprintf('+%d seconds', $delay / 1000));
+
+        $queryBuilder = $this->driverConnection->createQueryBuilder()
+            ->insert($this->configuration['table_name'])
+            ->values([
+                'body' => ':body',
+                'headers' => ':headers',
+                'queue_name' => ':queue_name',
+                'created_at' => ':created_at',
+                'available_at' => ':available_at',
+            ]);
+
+        $this->executeQuery($queryBuilder->getSQL(), [
+            ':body' => $body,
+            ':headers' => Type::getType(Type::JSON)->convertToDatabaseValue($headers, $this->driverConnection->getDatabasePlatform()),
+            ':queue_name' => $this->configuration['queue_name'],
+            ':created_at' => self::formatDateTime($now),
+            ':available_at' => self::formatDateTime($availableAt),
+        ]);
+    }
+
+    public function get(): ?array
+    {
+        $this->driverConnection->beginTransaction();
+        try {
+            $query = $this->driverConnection->createQueryBuilder()
+                ->select('m.*')
+                ->from($this->configuration['table_name'], 'm')
+                ->where('m.delivered_at is null OR m.delivered_at < :redeliver_limit')
+                ->andWhere('m.available_at <= :now')
+                ->andWhere('m.queue_name = :queue_name')
+                ->orderBy('available_at', 'ASC')
+                ->setMaxResults(1);
+
+            $now = \DateTime::createFromFormat('U.u', microtime(true));
+            $redeliverLimit = (clone $now)->modify(sprintf('-%d seconds', $this->configuration['redeliver_timeout']));
+            // use SELECT ... FOR UPDATE to lock table
+            $doctrineEnvelope = $this->executeQuery(
+                $query->getSQL().' '.$this->driverConnection->getDatabasePlatform()->getWriteLockSQL(),
+                [
+                    ':now' => self::formatDateTime($now),
+                    ':queue_name' => $this->configuration['queue_name'],
+                    ':redeliver_limit' => self::formatDateTime($redeliverLimit),
+                ]
+            )->fetch();
+
+            if (false === $doctrineEnvelope) {
+                $this->driverConnection->commit();
+
+                return null;
+            }
+
+            $doctrineEnvelope['headers'] = Type::getType(Type::JSON)->convertToPHPValue($doctrineEnvelope['headers'], $this->driverConnection->getDatabasePlatform());
```

Comment:
`Type::JSON` does not exists on early Doctrine versions [as the `low` test suite is highlighting](https://travis-ci.org/symfony/symfony/jobs/513749042). Therefore, I suggest to move it back to `STRING` (like the `body`) and use `json_encode`. Whoever wants to have the table with a `JSON` field can create it manually. (FYI @weaverryan)

Reply:
Yes i saw that. I was trying to require a minimum version for Doctrine but the Json type is available since v2.6.0 (tagged 2017-07-23 01:02) ... I've put back the json_encode / json_decode with the string type then